### PR TITLE
Torchgym

### PR DESCRIPTION
**Pull request by uellue**
Notebook showing how differentiable TemGym could work, including solving for rays with specific properties.

Note how the core parts don't rely on the model being linear. It just has to be differentiable in pytorch.

### CODE REVIEW

**Comment by uellue** [comment]
JAX looks quite good IMO! Since it is closer to NumPy and the documentation is very good, I'd be very happy to use this instead of PyTorch.

**Comment by uellue** [comment]
Ah, this could be relevant for implementation in TemGym: https://jax.readthedocs.io/en/latest/faq.html#how-to-use-jit-with-methods.

TL;DR it looks like making models immutable can help with JAX. 

Also nice: https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#jax-the-sharp-bits

> When walking about the countryside of Italy, the people will not hesitate to tell you that JAX has [“una anima di pura programmazione funzionale”](https://www.sscardapane.it/iaml-backup/jax-intro/).

**Comment by matbryan52** [comment]
Of course, some form of immutable representation of a given configuration can be generated from the existing code.

Bear in mind that the `(STEM)Model` is essentially the M of an MVC-ish GUI, hence why for `MicroscopeCalibration` it was necessary to change some parameters manually, as normally the GUI(Controller) would be used to do this. At the time the major goal was to revamp the GUI into something that separated the physics from the widgets and display, and `MicroscopeCalibration` had to use the same pieces to do what it needed to do, including using a stateful `Model`. Now that you have bigger plans we will of course try to find a way to express the same physics without a state, while maintaining the ability to run the GUI smoothly and without undue overhead.

**Comment by uellue** [comment]
It seems that this is what one would use in JAX to minimize a loss? https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.optimize.minimize.html

**Comment by matbryan52** [comment]
https://github.com/google-deepmind/optax

**Comment by matbryan52** [comment]
> Thank you for considering it, I hope it doesn't complicate things too much! I have a feeling that it will be well worth it.

So the "statefulness" part is easy to convert, 90% of the `Components` are essentially `dataclasses` anyway, so very simple to extract into `NamedTuples` if that is useful. Jax supports `NamedTuple` as a container class by default.

Then the `.step(self, rays: Rays) -> Generator[Rays, None, None]` method that exists on each class already can instead be a static `component_step(state: ComponentStateTuple, rays: Rays) -> Rays`, simply replacing `self` with the tuple, and any interior method calls with equivalent `static + state` methods. Almost all of the rest of the code in `components.py` and `model.py` exists to help setup components (i.e. they become a constructor of a `NamedTuple`), or to ensure the validity of a configuration, so essentially becomes a validator which will `raise` if you pass it in an invalid configuration (much as it does now when you try to set an invalid state). If you don't call the validator (formerly `Model`) then any set of component configuration would run but perhaps give unexpected results. You can already get this behavior in the existing code by doing :

```python
rays = source.get_rays(100)
rays = rays.propagate(rays, 0.1)
rays = next(lens.step(rays))
rays = rays.propagate(rays, 0.2)
rays = next(deflector.step(rays))
rays = rays.propagate(rays, 0.3)
image = detector.get_image(rays)
```
(note, we need the `next()`  because currently `step` is a generator since a `Component` can give us rays at intermediate locations, for example `DoubleDeflector` contains two `Deflector`, so we get access to the rays after each internal step).

The difficulty is the `Rays` object itself, which is what actually carries the computation through the model - right now it does a lot of heavy lifting to handle tracking the optical path length of a given ray, propagation, whether a ray gets blocked _en route_, current ray position, extra behavior when doing Gaussian beam propagation etc. The brute force method would be to just `copy` all these values to return a new `Rays` object (`NamedTuple` again, probably), but that's going to be a quite a substantial change. If all we cared about was the ray trajectories then we could just operate on `rays.data`, which is a `[5, n]` array, but then you lose other aspects.

I don't think it's insurmountable, just requires some thought to not put a big overhead on each component in the GUI.

**Comment by DavidLanders95** [comment]
I was thinking about this more yesterday, and something crossed my mind: 

Each operation you are applying with this model, including possible tilt and shift values for the descan error are always linear. That means the entire system will always be described by a 5x5 Ray Transfer Matrix. It would be a 3x3 in one dimension, assuming you are solving an equation of the form $y = ax + by + c$, and if you include two dimensions it becomes 5x5, if you ignore the constant offsets of c it becomes 4x4. Unless you write a function that is non linear in some sense, or you have tilt/shift errors coefficients that change as a function of position in a non linear way, each ray will see the same Ray Transfer Matrix. If each ray is not seeing the same 5x5 matrix, it means you have a very weird function describing the system, or your system is not linear and you need to use a larger transfer matrix to properly describe the system, i.e you have aberrations whereby I think you need a 19x19 Transfer matrix (or something of that size, can't remember exactly) to capture the behaviour. 

You might have already written this out by hand or solved it for this system before, but what I think this means is that we should never need to trace rays from the detector backwards and run an optimiser on them, and as long as we have the ray transfer matrix can always just invert it to find where rays should have come from. 

For instance, If we have a 2x2 ray transfer matrix for one dimension with no constant offsets, then

image

Once we have the [A, B, C, D] matrix, one can either use invert operation to find the coordinates and tilts from the detector that are needed to have rays leaves from a point source at the crossover.

If we invert this

$$
\begin{bmatrix}
x_0 \\
x_0'
\end{bmatrix}=
\frac{1}{AD - BC}
\begin{bmatrix}
D & -B \\
-C & A
\end{bmatrix}
\begin{bmatrix}
x_1 \\
x_1'
\end{bmatrix}.
$$

Solving for $x_0$

$$x_0 = \frac{D x_1 + B x_1'}{AD-BC} $$

since we know the $A, B, C, D$ values, and we know that $x_1$ is our detector coordinates and x_0 is 0, I think one can directly solve for $x_1'$ there and find all tilts without an optimiser?

And of course I am not saying at all that we don't need a differential model, and this is a simple place to start with a known solution - maybe you can check that pytorch gives you back this solution? 

The takeaways are that there is always a ray transfer matrix, in this system with 4DStem it will always be linear, even with scan and descan error. Even if you write a complicated scan and descan error function that is non-linear, locally each ray could see a linear ray transfer matrix, or probably a higher order transfer matrix where the coefficients are the aberration coefficients. Either way, one matrix or multiple linear matrices which are invertible will solve the system. 

Also it's worth mentioning that automatic differentiation is still useful in another way - it automatically will give us the numerical ray transfer matrix for the system, and saves us the hassle of calculating it manually. All one has to do is calculate the jacobian on the output rays with respect to the input rays, and the matrix that returns will be the ray transfer matrix. That's what I am setting up at the moment, also hopefully with the ability to differentiate with respect to parameters in the model also.

**Comment by uellue** [comment]
That's a good point, in particular about the Jacobian!

Since linear behavior is only the first approximation and many interesting things in a TEM cannot be described with such ray transfer matrices, TEMGYM should not be fundamentally limited to it, IMO. However, it is of course very interesting to easily extract a linear approximation around a given operating point!

Fully agreed on not needing to back-propagate. Instead we can just solve for the rays we are looking for using the forward direction and a solver.

All in all, it looks like having a model written with JAX is quite powerful!
* Solving for rays.
* Not limited to ray transfer matrices, supporting rather complex transfer functions.
* Very easy and efficient to extract approximate ray transfer matrices.

**Comment by DavidLanders95** [comment]
> That's a good point, in particular about the Jacobian!
> 
> Since linear behavior is only the first approximation and many interesting things in a TEM cannot be described with such ray transfer matrices, TEMGYM should not be fundamentally limited to it, IMO. However, it is of course very interesting to easily extract a linear approximation around a given operating point!
> 
> Fully agreed on not needing to back-propagate. Instead we can just solve for the rays we are looking for using the forward direction and a solver.
> 
> All in all, it looks like having a model written with JAX is quite powerful!
> 
> * Solving for rays.
> * Not limited to ray transfer matrices, supporting rather complex transfer functions.
> * Very easy and efficient to extract approximate ray transfer matrices.

As far as I am aware, I can't think of a case where a ray transfer matrix won't be present. As long as you can correctly calculate a high enough order of partial derivatives, at the end of the day aberration coefficients will just populate terms in the ray transfer matrix, and your output will be a function of not just x and x' to the power of 1, but x and x' to the power of 2, 3 and mixed terms also, The only time it won't work off the top of my head is with a biprism, but that still has a ray transfer matrix, just a different sign in a certain term for each side of the biprism. In the case of imaging that's a different story, but a ray transfer matrix can still tell you how to get a wavefield with amplitude and phase from image plane to detector if you know the ray transfer matrix of the system also. It's called the collins integral, and is like fresnel diffraction integral but with ABCD terms instead of distances between components and phase masks.

**Comment by uellue** [comment]
The biprism is a good example, but also the specimen and apertures behave in a way that cannot be described well with a transfer matrix, even if it includes higher orders. One can express multislice as a higher-dimensional transfer matrix, see for example https://arxiv.org/abs/2205.03902, but that becomes gigantic and rather cumbersome IMO. It is much better to just write it down as a sequence of operations, as it is usually done. From a user's perspective it will also be nice to be able to express a transfer function without needing to calculate a higher order matrix, even if that might be possible.

By the way, how would one express things like spiral distortion? I guess that should work similar to the rotation, but with a polynomial to get rotation as a function of distance from the optical axis, right?

I remember playing around with polynomials for ray transfer, and the required order to remain numerically correct exploded when several such elements are chained. So I guess it will be quite, quite practical to model the column with "any function, as long as it works with JAX" and then extract a polynomial approximation when needed or desired. :-)

I'm still not entirely clear when and where such an approximation would be needed. I can imagine the following ones:
* Calculate many similar rays very quickly in cases where the function is smooth so that fast interpolation works
* Abstract away "boring" parts of the column that have many optical elements, but can be described well with a simple transfer function.
* Extract optical properties (focal length, lens errors and such) from ray propagation in electromagnetic fields or other non-trivial functions, such as phase plates or zone plates.

**Comment by DavidLanders95** [comment]
Yes, finding the edges of apertures or how to move them to get them to the optical axis is one useful part, but remember the ABCD matrix for that system will still exist, and can still be inverted, one just needs to just propagate to the aperture. Remove rays that hit the edges, and continue with the ABCD matrix to the end. Samples are essentially the same as an aperture in that regard. The existence or not of the aperture or sample will not change the ABCD matrix for the system (obviously I guess!). If we propagate forward and remove rays at the aperture, the rays that are left behind will still carry an invertible transfer matrix. 

The output of a sample can still be propagated with an ABCD matrix also as I mentioned earlier: https://opg.optica.org/josaa/fulltext.cfm?uri=josaa-41-12-2361&id=563424 but it requires an FFT to solve it (one could of course trace all the rays that leave the sample and interfere them, but the FFT is much more efficient at solving that!) The paper has some discussion on how to include apertures in that analysis too but I have not understood it enough to know if it's useful. Either way just propagate to the aperture or sample, apply the effect, and propagate forward again. 

You are right, spiral distortion is a mixed aberration term, where the output position in the object plane is a function of input position^3, composed of both x and y terms. 

image
I am fairly certain it is these terms here, whose coefficients are zero for an electrostatic lens, but not 0 for a magnetic lens,

If we write the input matrix in jax of the rays not just of [x, x', y, y', 1], but of [x, x', y, y'... higher order terms... such as x'3 etc ] when we calculate the jacobian, and probably the hessian and higher order derivatives, we should get the transfer matrix with aberration coefficients inside it for the entire system. We are essentially using automatic differentiation to do a taylor expansion of the system, and the aberration coefficients pop out. It should not be a problem for autodiff to handle this numerically, since it solves derivatives algebraicially, and not with finite differences.
